User updates RSS feed

rascalking opened the pull request:
This addresses issue #4603.

yeraydiazdiaz commented on warehouse/templates/rss/user_updates.xml:
```diff
@@ -0,0 +1,13 @@
+{% extends "base.xml" %}
+{% block title %}{{ user.name }} recent updates{% endblock %}
```
This should be `username` it's currently not rendering anything resulting `<title> recent updates</title>`

```suggestion
{% block title %}{{ user.username }} recent updates{% endblock %}
```

Also happening in the line below.

rascalking replied:
Fixed.

rascalking commented on warehouse/templates/accounts/profile.html:
```diff
@@ -39,6 +39,12 @@ <h1 class="author-profile__name">{{ user.name }}</h1>
+              <i class="fa fa-rss" aria-hidden="true"></i>
+              <span class="sr-only">{% trans %}RSS Feed{% endtrans %}</span>
+              &nbsp;&nbsp;<a href="{{ request.route_path('rss.user_updates', username=user.username) }}">{% trans %}Recent updates{% endtrans %}</a></p>
+            {% endif %}
```
I think adding in the extra `alternate` link probably makes sense, regardless.  I'll add it in with the bugfix.

di replied:
I think it makes sense to make the link visible in the profile somewhere, otherwise it would be challenging to find these feeds. The other site-wide ones are listed at https://warehouse.readthedocs.io/api-reference/feeds/, and this PR should probably update that documentation as well.

rascalking replied:
Done.

di commented on warehouse/rss/views.py:
```diff
@@ -72,3 +73,34 @@ def rss_packages(request):
+            1 * 24 * 60 * 60,  # 1 day
+            stale_while_revalidate=1 * 24 * 60 * 60,  # 1 day
+            stale_if_error=5 * 24 * 60 * 60,  # 5 days
+            keys=["all-projects"],
```
This line would cause the cache for every user RSS feed to be purged whenever any new project was created. It's not necessary since we already have the `context=User` line above.

rascalking replied:
Thanks for the catch.  I just copied and pasted this from the other RSS views, should have read through the code to understand it first.

di commented on warehouse/rss/views.py:
```diff
@@ -72,3 +73,33 @@ def rss_packages(request):
+    latest_releases = (
+        request.db.query(Release)
+        .join(Project)
+        .filter(Project.users.any(username=user.username))
```
Note that this doesn't mean that this user has made this release, but rather that this user has some role on a project that has made a release.

In #4603 the request is for both:

> (I actually want 2 feeds: one for "stuff I did directly" and one for "projects I am a maintainer on that did releases", to account for projects where the actual releases are done by a bot-user, or by a release manager.)

I think it's fine if we're only implementing half of #4603 here, but we might want to think a little more about what we're calling this feed since `updates` is pretty generic and we might have multiple feeds here eventually.

We also might want to make it more clear on the profile page what this feed is exactly.

rascalking replied:
Hrm...maybe call them "My Releases" vs. "My Projects' Releases"?

di commented on tests/unit/test_routes.py:
```diff
@@ -283,6 +283,13 @@ def add_policy(name, filename):
         pretend.call("rss.packages", "/rss/packages.xml", domain=warehouse),
+        pretend.call(
+            "rss.user_updates",
+            "/user/{username}/updates.xml",
```
Looking at #7013, I think this should probably be `/rss/user/{username}/something.xml`, where `something` is distinctly one of the two types of feeds discussed in #4603.

rascalking commented:
Okay, I implemented both feeds, named them `My Releases` and `My Projects' Releases`, and updated the urls as suggested.  I think this is all set, let me know if there's anything I missed.

di commented:
Hi @rascalking, can you update the merge conflicts here now that #7013 has been merged? You might also want to make some updates (docs, etc.) to align more with what was merged there.

Also, after giving it some thought I think the URLs should be:

* Uploaded Releases: `https://pypi.org/rss/user/<username>/uploaded_releases.xml`
* Maintained Releases: `https://pypi.org/rss/user/<username>/maintained_releases.xml`

rascalking commented:
Sorry, I don't think I'm going to be able to finish this.
